File: gui/menu/workspace/workspace_describe.cpp
#include "workspace_describe.h"


MenuWorkspaceDescribe::MenuWorkspaceDescribe(QWidget* parent) : 
    QIodeSettings(parent)
{
    setupUi(this);

    QList<QString> qIodeTypes;
    for(const std::string& iode_type: vIodeTypes) qIodeTypes << QString::fromStdString(iode_type);

    wCommentsDescription = new WrapperQLineEdit(label_comments->text(), *lineEdit_comments, OPTIONAL_FIELD);
    wEquationsDescription = new WrapperQLineEdit(label_equations->text(), *lineEdit_equations, OPTIONAL_FIELD);
    wIdentitiesDescription = new WrapperQLineEdit(label_identities->text(), *lineEdit_identities, OPTIONAL_FIELD);
    wListsDescription = new WrapperQLineEdit(label_lists->text(), *lineEdit_lists, OPTIONAL_FIELD);
    wScalarsDescription = new WrapperQLineEdit(label_scalars->text(), *lineEdit_scalars, OPTIONAL_FIELD);
    wTablesDescription = new WrapperQLineEdit(label_tables->text(), *lineEdit_tables, OPTIONAL_FIELD);
    wVariablesDescription = new WrapperQLineEdit(label_variables->text(), *lineEdit_variables, OPTIONAL_FIELD);

    wCommentsDescription->setQValue(QString::fromStdString(KDBComments().get_description()));
    wEquationsDescription->setQValue(QString::fromStdString(KDBEquations().get_description()));
    wIdentitiesDescription->setQValue(QString::fromStdString(KDBIdentities().get_description()));
    wListsDescription->setQValue(QString::fromStdString(KDBLists().get_description()));
    wScalarsDescription->setQValue(QString::fromStdString(KDBScalars().get_description()));
    wTablesDescription->setQValue(QString::fromStdString(KDBTables().get_description()));
    wVariablesDescription->setQValue(QString::fromStdString(KDBVariables().get_description()));

    className = "MENU_WORKSPACE_DESCRIBE";
    loadSettings();
}

MenuWorkspaceDescribe::~MenuWorkspaceDescribe()
{
    delete wCommentsDescription;
    delete wEquationsDescription;
    delete wIdentitiesDescription;
    delete wListsDescription;
    delete wScalarsDescription;
    delete wTablesDescription;
    delete wVariablesDescription;
}

void MenuWorkspaceDescribe::update_descriptions()
{
    try
    {
        std::string comments_description = wCommentsDescription->extractAndVerify().toStdString();
        std::string equations_description = wEquationsDescription->extractAndVerify().toStdString();
        std::string identites_description = wIdentitiesDescription->extractAndVerify().toStdString();
        std::string lists_description = wListsDescription->extractAndVerify().toStdString();
        std::string scalars_description = wScalarsDescription->extractAndVerify().toStdString();
        std::string tables_description = wTablesDescription->extractAndVerify().toStdString();
        std::string variables_description = wVariablesDescription->extractAndVerify().toStdString();

        if (!comments_description.empty())
        {
            KDBComments kdb;
            kdb.set_description(comments_description);
        }

        if (!equations_description.empty())
        {
            KDBEquations kdb;
            kdb.set_description(equations_description);
        }

        if (!identites_description.empty())
        {
            KDBIdentities kdb;
            kdb.set_description(identites_description);
        }

        if (!lists_description.empty())
        {
            KDBLists kdb;
            kdb.set_description(lists_description);
        }

        if (!scalars_description.empty())
        {
            KDBScalars kdb;
            kdb.set_description(scalars_description);
        }

        if (!tables_description.empty())
        {
            KDBTables kdb;
            kdb.set_description(tables_description);
        }

        if (!variables_description.empty())
        {
            KDBVariables kdb;
            kdb.set_description(variables_description);
        }
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }

    this->accept();
}

void MenuWorkspaceDescribe::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/menu/workspace/workspace_describe.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from settings import MixinSettingsDialog
from .ui_workspace_describe import Ui_MenuWorkspaceDescribe


from iode import (comments, equations, identities, lists, 
                  scalars, tables, variables)


class MenuWorkspaceDescribe(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuWorkspaceDescribe()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)
        self.load_settings()

    @Slot()
    def update_descriptions(self):
        try:
            # Extract and verify the descriptions
            comments_description: str = self.ui.lineEdit_comments.text()
            equations_description: str = self.ui.lineEdit_equations.text()
            identities_description: str = self.ui.lineEdit_identities.text()
            lists_description: str = self.ui.lineEdit_lists.text()
            scalars_description: str = self.ui.lineEdit_scalars.text()
            tables_description: str = self.ui.lineEdit_tables.text()
            variables_description: str = self.ui.lineEdit_variables.text()

            # Update the descriptions if they're not empty
            if comments_description:
                comments.description = comments_description
            if equations_description:
                equations.description = equations_description
            if identities_description:
                identities.description = identities_description
            if lists_description:
                lists.description = lists_description
            if scalars_description:
                scalars.description = scalars_description
            if tables_description:
                tables.description = tables_description
            if variables_description:
                variables.description = variables_description

            self.accept()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", "Failed to updated the description "
                                "of the IODE database(s)\n" + str(e))
